paint/PAINT_ColourPicker.cpp

```diff
@@ -77,34 +109,74 @@ void ColourPicker::setPrimarySecondaryFromActiveDraw() const
 	}
 }
 
+
+void ColourPicker::swapActiveColour() const
+{
+	primaryColourDisplay_->swapIsActive();
+	secondaryColourDisplay_->swapIsActive();
+
+	setDrawColourFromActive();
+	//update();
+
+	//rerender();
+}
+
+void ColourPicker::updateColourDisplaysFromDrawWindow() const
+{
+	const auto primaryRGBA = drawWindowPtr_->getPrimaryRGBA();
+	const auto secondaryRGBA = drawWindowPtr_->getSecondaryRGBA();
+	if (!swappedDisplays_) {
+		//primaryColourDisplay_->setColour(drawWindowPtr_->getPrimaryColour());
+		//secondaryColourDisplay_->setColour(drawWindowPtr_->getSecondaryColour());
+		primaryColourDisplay_->setColour(gfx::Colour(primaryRGBA[0], primaryRGBA[1], primaryRGBA[2], primaryRGBA[3]));
+		secondaryColourDisplay_->setColour(gfx::Colour(secondaryRGBA[0], secondaryRGBA[1], secondaryRGBA[2], secondaryRGBA[3]));
+	}
+	else {
+		//primaryColourDisplay_->setColour(drawWindowPtr_->getSecondaryColour());
+		//secondaryColourDisplay_->setColour(drawWindowPtr_->getPrimaryColour());
+		primaryColourDisplay_->setColour(gfx::Colour(secondaryRGBA[0], secondaryRGBA[1], secondaryRGBA[2], secondaryRGBA[3]));
+		secondaryColourDisplay_->setColour(gfx::Colour(primaryRGBA[0], primaryRGBA[1], primaryRGBA[2], primaryRGBA[3]));
+	}
+}
+
+void ColourPicker::updateColourDisplays() const
+{
+	setPrimarySecondaryFromActiveDraw();
+	primaryColourDisplay_->draw();
+	secondaryColourDisplay_->draw();
+	SDL_RenderPresent(renderer_);
```

Comment:
Does `SDL_RenderPresent(...)` get called from the main render loop? If so then this is unnecessary. If not it probably ought to as this forces us to repaint everything several times per frame. If all else fails, use a dirty flag at the end of the main loop (but before the repaint) that tells SDL to redraw itself. But I think the design was a continuous update / draw cycle.
**MUST**

Reply:
removed SDL_RenderPreset(...), you are correct it should not be called and should have same purpose as other 'update' functions.